service/history/task/cross_cluster_task.go

```diff
@@ -208,47 +265,396 @@ func newCrossClusterTaskBase(
 			GetCrossClusterTaskMetricsScope(taskInfo.GetTaskType()), taskInfo.GetDomainID(),
 			logger,
 		),
-		resubmitTaskFn: resubmitTaskFn,
-		maxRetryCount:  maxRetryCount(),
+		taskExecutor:  taskExecutor,
+		taskProcessor: taskProcessor,
+		redispatchFn:  redispatchFn,
+		maxRetryCount: maxRetryCount(),
 	}
 }
 
 // cross cluster source task methods
 
 func (t *crossClusterSourceTask) Execute() error {
-	panic("Not implement")
+	executionStartTime := t.timeSource.Now()
+
+	defer func() {
+		t.scope.IncCounter(metrics.TaskRequestsPerDomain)
+		t.scope.RecordTimer(metrics.TaskProcessingLatencyPerDomain, time.Since(executionStartTime))
+	}()
+
+	logEvent(t.eventLogger, "Executing task")
+	return t.taskExecutor.Execute(t, true)
 }
 
 func (t *crossClusterSourceTask) Ack() {
-	panic("Not implement")
+	// do not set t.state to Acked, which will prevent the task from being fetched again
+	// state and processingState will be updated by the task executor
+	t.Lock()
+	state := t.state
+	processingState := t.processingState
+	t.Unlock()
+
+	logEvent(t.eventLogger, "executed task", "processing state", processingState)
+
+	if state == ctask.TaskStateAcked {
+		t.scope.RecordTimer(metrics.TaskAttemptTimerPerDomain, time.Duration(t.attempt))
+		t.scope.RecordTimer(metrics.TaskLatencyPerDomain, time.Since(t.submitTime))
+		t.scope.RecordTimer(metrics.TaskQueueLatencyPerDomain, time.Since(t.GetVisibilityTimestamp()))
+
+		if t.eventLogger != nil && t.attempt != 0 {
+			// only dump events when the task has been retried
+			t.eventLogger.FlushEvents("Task processing events")
+		}
+	}
 }
 
 func (t *crossClusterSourceTask) Nack() {
-	panic("Not implement")
+	// Nack will be called when we are unable to move the next step for processing
+	// basically add the task to redispatch queue, so it can be retried later.
+	// wether the task state is setting to Nacked or not is not important, since
+	// task will be retried forever
+	logEvent(t.eventLogger, "Nacked task")
```

Comment:
Are we going to have non-retryable errors? We talked about domains but as an example, workflow id might not exist; there could be some other cases we might know that retry will fail.